### ISSUE TITLE
Recognition of Special Characters

### ISSUE THREAD

**Issue by rayush7** [bgshih/crnn#60]
@bgshih I am able to train the model on a dataset consisting of alphabets (a-z, A-Z) and digits (0-9) but I also want to train the model on special characters like $, #, @. The images in my dataset have labels like "Subtotal=$30". I made following changes to the ascii2label function and label2ascii function in utilities.lua.

Modified ascii2label function:

 if ascii >= 48 and ascii <= 57 then -- '0'-'9' are mapped to 1-10
            label = ascii - 47
        elseif ascii >= 65 and ascii <= 90 then -- 'A'-'Z' are mapped to 11-36
            label = ascii - 64 + 10
        elseif ascii >= 97 and ascii <= 122 then -- 'a'-'z' are mapped to 11-36
            label = ascii - 96 + 10
	   -- Adding Special Symbols here
        elseif ascii >=32 and ascii<=47 then  -- (space) - / are mapped to 37-52
	    	label = ascii - 31 + 36
        elseif ascii >=58 and ascii<=64 then -- : - @ are mapped to 53-59
	    	label = ascii - 57+52
        elseif ascii >=91 and ascii <= 96 then -- [ - are mapped to 60-65
	    	label = ascii - 90 + 59
        elseif ascii>=123 and ascii <= 127 then -- { - are mapped to 66-70
	    	label = ascii - 122 + 65
        end

Modified label2ascii() function

if label >= 1 and label <= 10 then
            ascii = label - 1 + 48
        elseif label >= 11 and label <= 36 then
            ascii = label - 11 + 97
        elseif label == 0 then -- used when displaying raw predictions
            ascii = string.byte('-')
        -- For special symbols
        elseif label >=37 and label <=52 then
            ascii = label - 37 + 32
        elseif label >=53 and label <=59 then
            ascii = label - 53 + 58
        elseif label >= 60 and label <=65 then
            ascii = label - 60 + 91
        elseif label >= 66 and label <=70 then
            ascii = label - 66 + 123
        end

I am getting the following error in training.lua in the line
local loss = criterion:forward(output, target, true) / nFrame

Error I got
terminate called after throwing an instance of 'std::invalid_argument'
what():  index out of range
Aborted (core dumped)

and sometimes this

terminate called recursively
terminate called after throwing an instance of 'std::invalid_argument'
terminate called recursively
Aborted (core dumped)

Please help where I am making mistake. Is there anything else I need to do for incorporating special symbols in crnn?

**Comment by rayush7** [comment]
I was able to resolve the issue. Actually in the config.lua file, which will be used for training the model, there is a parameter called nClasses. Initially it is set to 36 corresponding to a-z or A-Z and 0-9. One needs to change that parameter according to the number of classes in one's case. For example for I have 70 classes in total including alphabets, digits and special symbols. So I set nClasses=70 in config.lua and it works!
